swarm/storage/mru/handler.go

```diff
@@ -0,0 +1,509 @@
+	// to update
+
+	var r SignedResourceUpdate
+	if err := r.parseUpdateChunk(chunkAddr, data); err != nil {
+		log.Warn("Invalid resource chunk")
+		return false
+	}
+
+	return true
+}
+
+// GetContent retrieves the data payload of the last synced update of the Mutable Resource
+func (h *Handler) GetContent(rootAddr storage.Address) (storage.Address, []byte, error) {
+	rsrc := h.get(rootAddr)
+	if rsrc == nil || !rsrc.isSynced() {
+		return nil, nil, NewError(ErrNotFound, " does not exist or is not synced")
+	}
+	return rsrc.lastKey, rsrc.data, nil
+}
+
+// GetLastPeriod retrieves the period of the last synced update of the Mutable Resource
+func (h *Handler) GetLastPeriod(rootAddr storage.Address) (uint32, error) {
+	rsrc := h.get(rootAddr)
+	if rsrc == nil {
+		return 0, NewError(ErrNotFound, " does not exist")
+	} else if !rsrc.isSynced() {
+		return 0, NewError(ErrNotSynced, " is not synced")
+	}
+	return rsrc.period, nil
+}
+
+// GetVersion retrieves the period of the last synced update of the Mutable Resource
+func (h *Handler) GetVersion(rootAddr storage.Address) (uint32, error) {
+	rsrc := h.get(rootAddr)
+	if rsrc == nil {
+		return 0, NewError(ErrNotFound, " does not exist")
+	} else if !rsrc.isSynced() {
+		return 0, NewError(ErrNotSynced, " is not synced")
+	}
+	return rsrc.version, nil
+}
+
+// \TODO should be hashsize * branches from the chosen chunker, implement with FileStore
+func (h *Handler) chunkSize() int64 {
+	return chunkSize
+}
+
+// New creates a new metadata chunk out of the request passed in.
+func (h *Handler) New(ctx context.Context, request *Request) error {
+
+	// frequency 0 is invalid
+	if request.frequency == 0 {
+		return NewError(ErrInvalidValue, "frequency cannot be 0 when creating a resource")
+	}
+
+	// make sure name only contains ascii values
+	if !isSafeName(request.name) {
+		return NewError(ErrInvalidValue, fmt.Sprintf("invalid name: '%s'", request.name))
+	}
+
+	// make sure owner is set to something
+	var zeroAddr = common.Address{}
+	if request.ownerAddr == zeroAddr {
+		return NewError(ErrInvalidValue, "ownerAddr must be set to create a new metadata chunk")
+	}
+
+	// create the meta chunk and store it in swarm
+	chunk, metaHash := request.resourceMetadata.newChunk()
+	if request.metaHash != nil && !bytes.Equal(request.metaHash, metaHash) ||
+		request.rootAddr != nil && !bytes.Equal(request.rootAddr, chunk.Addr) {
+		return NewError(ErrInvalidValue, "metaHash in UpdateRequest does not match actual metadata")
+	}
+
+	request.metaHash = metaHash
+	request.rootAddr = chunk.Addr
+
+	h.chunkStore.Put(chunk)
+	log.Debug("new resource", "name", request.name, "startBlock", request.startTime, "frequency", request.frequency, "owner", request.ownerAddr)
+
+	// create the internal index for the resource and populate it with the data of the first version
+	rsrc := &resource{
+		resourceUpdate: resourceUpdate{
+			updateHeader: updateHeader{
+				UpdateLookup: UpdateLookup{
```

Comment:
This thing makes me uncomfortable. But let's save the discussion for the last commit.

Reply:
Yes, I know it looks a bit ugly, but you will see why it makes sense to break it down and the reusability and clarity it provides when it is time to make a change.

In any case, this I believe was finally rewritten as:
```
rsrc:= new(resource)
rsrc.rootAddr = request.rootAddr
rsrc.resourceMetadata = request.metadata
rsrc.updated = time.Now()
```

This said, I think we could potentially remove one level, but I also think we should revisit the `resource` struct, because after spending quite a bit of time on this code I think it might not make sense to cache anything other than the resource metadata.